### PR TITLE
[QDP] feat: add credit card fraud benchmark + amplitude encoding optimizations

### PR DESCRIPTION
**Pull request by rich7420**
### Changes
  - **New benchmark**: `encoding_benchmarks/qdp_pipeline/creditcardfraud_amplitude.py` — 5-qubit
  amplitude VQC on Credit Card Fraud data, aligned with PennyLane baseline (same circuit, loss,
  optimizer). Closes the QDP vs baseline training time gap from ~22% slower to <1% gap.
  - **New baseline**: `encoding_benchmarks/pennylane_baseline/creditcardfraud_amplitude.py` —
  PennyLane reference implementation with AUPRC/F1 metrics for imbalanced data.
  - **`QuantumDataLoader` API**: added `source_array(X)` (in-memory, no temp file),
  `as_torch(device)`, and `as_numpy()` for ergonomic batch output format.
  - **Rust `PipelineIterator`**: added `new_from_array()` constructor; `InMemory` `next_batch` now
   passes `&data[start..end]` slice directly (no per-batch `to_vec()`).
  - **`amplitude.rs`**: moved D2H norm validation to after encode kernel + `device.synchronize()`,
   eliminating a mid-pipeline GPU→CPU roundtrip in `encode_batch`.
  - **Bug fixes** (iris + creditcard benchmarks): `requires_grad=False` on all data arrays to
  prevent AdamOptimizer from computing unnecessary gradients through state vectors;
  `AmplitudeEmbedding(normalize=False)` in place of `StatePrep`; `.real` extraction after
  `torch.from_dlpack()` to convert complex128 DLPack output to float64.

### Motivation

The existing QDP benchmark suite only covers the Iris dataset (100 samples, 2 qubits), which is too small to surface real-world data-loading and encoding bottlenecks. Credit Card Fraud (284,807 transactions, 5 qubits) is a standard imbalanced-classification benchmark from Kaggle/OpenML that stresses the full QDP pipeline — batch iteration, GPU encoding, and training — at realistic scale.

Adding this benchmark serves this purposes:
- **Expands loader API coverage**: `source_array(X)`, `as_torch()`, and `as_numpy()` are exercised end-to-end by the new benchmark and tests, catching integration issues across the Python → PyO3 → Rust → CUDA boundary.

### Checklist

- [ ] Bug fix
- [x] New feature
- [x] Refactoring
- [ ] Documentation
- [ ] Test
- [ ] CI/CD pipeline
- [ ] Other
- [x] Added or updated unit tests for all changes
- [x] Added or updated documentation for all changes

### CODE REVIEW

**Comment by rich7420** [comment]
It seems a little be too big one. Sorry about that.

File: qdp/qdp-python/qumat_qdp/loader.py [hunk 1]
```diff
@@ -254,6 +343,25 @@ def _create_iterator(self) -> Iterator[object]:
...
+            return
+        kind = self._output_format[0]
+        if kind == "torch":
+            device = self._output_format[1]
+            for qt in raw_iter:
+                t = _torch.from_dlpack(qt)
+                yield t.cpu() if device == "cpu" else t
+        elif kind == "numpy":
+            for qt in raw_iter:
+                yield _torch.from_dlpack(qt).cpu().numpy()
```
Comment (viiccwen) [inline qdp/qdp-python/qumat_qdp/loader.py]: `as_torch()` validates that `torch` is installed, but `as_numpy()` does not. Then `_wrap_iterator()` calls `_torch.from_dlpack(...)` for the `"numpy"` path.

Does it mean `as_numpy()` can succeed at configuration time and then fail during iteration with an unclear runtime error if PyTorch is not installed. 🤔

**Review by viiccwen — COMMENTED** [review]
Thx for contributing! 🙌
I'll look deeper tomorrow, and I think we should add tests to cover new loader APIs, especially the new behavior crosses Python, PyO3, Rust, and CUDA boundaries.

**Comment by ryankert01** [comment]
> amplitude.rs: moved D2H norm validation to after encode kernel + device.synchronize(),
eliminating a mid-pipeline GPU→CPU roundtrip in encode_batch.

nice

File: qdp/qdp-python/qumat_qdp/loader.py (hunk 1 above)
Comment (rich7420) [reply qdp/qdp-python/qumat_qdp/loader.py]: oh , nice catch! you're right

**Comment by 400Ping** [comment]
Please solve conflicts

**Comment by rich7420** [comment]
plz take a look and test, not hurry once you have time

**Review by ryankert01 — COMMENTED** [review]
The pennylane baseline uses CPU for training, whereas qdp pipeline has dual CPU/GPU. We can simplify both to use GPU for training always.

**Comment by rich7420** [comment]
no problem!

**Comment by 400Ping** [comment]
Please solve conflicts

**Comment by 400Ping** [comment]
Can you also give some more context in the PR description for why is this needed?

**Review by 400Ping — CHANGES_REQUESTED** [review]
The array-loader optimization claim does not match the implementation in this PR.
`create_array_loader()` says batching uses slices without per-batch `to_vec()`, but
`PipelineIterator::take_batch_from_source()` still clones each in-memory batch with
`data[start..end].to_vec()`.

**Comment by rich7420** [comment]
> The array-loader optimization claim does not match the implementation in this PR. `create_array_loader()` says batching uses slices without per-batch `to_vec()`, but `PipelineIterator::take_batch_from_source()` still clones each in-memory batch with `data[start..end].to_vec()`.

Great catch! You are right that the code in take_batch_from_source() was misleading.
Actually, next_batch() already handles the InMemory variant directly using the zero-copy data[start..end] slice logic and never calls take_batch_from_source() for it. To avoid confusion and make the implementation match the documentation claims, I've replaced the dead-code InMemory arm in take_batch_from_source() with unreachable!().
I've also rebased the branch onto main and resolved all the merge conflicts. Thanks for the review!

**Comment by guan404ming** [comment]
> Great catch! You are right that the code in take_batch_from_source() was misleading.
Actually, next_batch() already handles the InMemory variant directly using the zero-copy data[start..end] slice logic and never calls take_batch_from_source() for it. To avoid confusion and make the implementation match the documentation claims, I've replaced the dead-code InMemory arm in take_batch_from_source() with unreachable!().
I've also rebased the branch onto main and resolved all the merge conflicts. Thanks for the review!

Make sense to me, thanks for the update. I could code change wise looks nie, plz help resolve the conflict, thanks!

**Comment by 400Ping** [comment]
@rich7420 Are you planning to continue to work on this ? If so please rebase and solve conflicts.
